README.md
README.md
================

# seedr

![](seeds.png)

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/efernandezpascual/seedr.svg?branch=master)](https://travis-ci.com/efernandezpascual/seedr)
<!-- badges: end -->

## Overview

`seedr` is an R package that provides functions to fit hydro and thermal
time germination models. These models characterize seed lots by two sets
of parameters: (i) the physiological thresholds (water, temperature)
between which the seed lot can germinate, and (ii) the
physiological-time units that the seed lot needs to accumulate before it
can germinate. `seedr` fits the hydro time model of Bradford (Gummerson
1986, Bradford 1990, Bewley et al. 2013) and the thermal time model of
Garcia-Huidobro (Garcia-Huidobro et al. 1982, Gummerson 1986, Bewley et
al. 2013). It allows to fit models to grouped datasets, i.e. datasets
containing multiple species, seedlots or experiments.

## Installation

``` r
# Install stable version from CRAN
install.packages("seedr")

# Install last development version from GitHub
install.packages("devtools") # Install devtools if you don't have it
devtools::install_github("efernandezpascual/seedr") # Get seedr from GitHub
```

## Usage

``` r
library(seedr)
# seedr relies on physiotime(), a wrapper function that formats your data and fits a hydro/thermal time model
physiotime(grasses, # germination dataset
           x = "psi", # experimental treatment
           method = "bradford", # hydrotime model
           groups = c("species")) # group dataset by species
```

    ## A list of physiological time germination models 
    ##  calculated for the following 2 groups: 
    ##  
    ## species: Anisantha rubens 
    ## Bradford's hydrotime model 
    ##  Water potential levels in experiment: -1.6 -1 -0.8 -0.6 -0.4 -0.2 0 
    ##  Theta - Hydrotime constant: 50.54 
    ##  Psib50 - Base water potential (median): -1.59 
    ##  Sigma of the base water potential: 0.4 
    ##  R2: 0.94 
    ##  
    ## species: Bromus hordeaceus 
    ## Bradford's hydrotime model 
    ##  Water potential levels in experiment: -1.6 -1 -0.8 -0.6 -0.4 -0.2 0 
    ##  Theta - Hydrotime constant: 59.81 
    ##  Psib50 - Base water potential (median): -1.75 
    ##  Sigma of the base water potential: 0.38 
    ##  R2: 0.86 
    ## 

``` r
# for more information read the seedr vignette
vignette("seedr")
```

## References

  - Bewley, J. D., Bradford, K. J., Hilhorst, H. W., & Nonogaki, H.
    (2013). Environmental Control of Germination. In *Seeds: Physiology
    of Development, Germination and Dormancy, 3rd Edition*
    (pp. 302-317). Springer, New York, NY.
  - Bradford, K. J. (1990). A water relations analysis of seed
    germination rates. *Plant Physiology*, 94(2), 840-849.
  - Garcia-Huidobro, J., Monteith, J. L., & Squire, G. R. (1982). Time,
    temperature and germination of pearl millet (*Pennisetum typhoides*
    S. & H.) I. Constant temperature. *Journal of Experimental Botany*,
    33(2), 288-296.
  - Gummerson, R. J. (1986). The effect of constant temperatures and
    osmotic potentials on the germination of sugar beet. *Journal of
    Experimental Botany*, 37(6), 729-741.


NEWS.md
## version 0.3.0.9000

## version 0.3.0

This is the first public release.


R/bradford.R
#' Fits Bradford's hydrotime model
#'
#' \code{bradford} fits a hydrotime seed germination model using the method of
#' Bradford (Gummerson 1986, Bradford 1990, Bewley et al. 2013). This function
#' can be used only with one-group dataset, i.e. one seed lot of one species. To
#' fit models to grouped datasets (multi-seedlots, multi-species) use the
#' function \code{physiotime} instead.
#'
#' @usage bradford(d)
#' @param d a data.table within a "physiodata" object, containing the cumulative
#'   germination proportion at each scoring time and water potential treatment.
#' @return \code{bradford} returns a S3 object of class "bradford" with the
#'   results of fitting the hydrotime model. The generic functions
#'   \code{summary} and \code{plot} are used to obtain and visualize the model
#'   results.
#' @examples
#' # format dataset with physiodata
#' anisantha <- physiodata(subset(grasses, species == "Anisantha rubens"), x = "psi")
#' # bradford() uses the $proportions element within the physiodata object
#' b <- bradford(anisantha$proportions)
#' b # prints the main hydrotime variables
#' summary(b) # returns the main hydrotime variables as a data.table
#' plot(b) # plots the fitted model
#' @references Bewley, J. D., Bradford, K. J., Hilhorst, H. W., & Nonogaki, H.
#'   (2013). Hydrotime Model of Germination. In Seeds: Physiology of
#'   Development, Germination and Dormancy, 3rd Edition (pp. 303-307). Springer,
#'   New York, NY.
#'
#'   Bradford, K. J. (1990). A water relations analysis of seed germination
#'   rates. Plant Physiology, 94(2), 840-849.
#'
#'   Gummerson, R. J. (1986). The effect of constant temperatures and osmotic
#'   potentials on the germination of sugar beet. Journal of Experimental
#'   Botany, 37(6), 729-741.
#' @export
bradford <- function(d) tryCatch(
  {
    d <- d[! germination.mean %in% c(0, 1)]
    d[, probit := qnorm(germination.mean, 0, 1)]
    theta <- bradtheta(d[, probit], d[, treatment], 1/d[, time])
    d[, psibg := treatment - theta / time]

    hlm <- lm(probit ~ psibg, data = d)
    b <- summary(hlm)$coefficients[1, 1]
    m <- summary(hlm)$coefficients[2, 1]

    z <- list()

    z$data <- d

    z$parameters$levels <- d[, .(Levels = unique(treatment))][, Levels]
    z$parameters$theta <- theta
    z$parameters$psib50 <- -b/m
    z$parameters$sigma <- 1/m
    z$parameters$R2 <- summary(hlm)$r.squared

    class(z) <- "bradford"
    z
  }, error = function(e) e)

# bradford generic functions

#' @export
print.bradford <- function(x, ...)
{
  cat("Bradford's hydrotime model", "\n",
      "Water potential levels in experiment:",
      round(x$parameters$levels, 1), "\n",
      "Theta - Hydrotime constant:",
      round(x$parameters$theta, 2), "\n",
      "Psib50 - Base water potential (median):",
      round((x$parameters$psib50), 2), "\n",
      "Sigma of the base water potential:",
      round(x$parameters$sigma, 2), "\n",
      "R2:", round(x$parameters$R2, 2), "\n", "\n")
}

#' @export
summary.bradford <- function(object, ...)
{
  data.table(
    method = class(object),
    n.treatments = length(object$parameters$levels),
    theta = object$parameters$theta,
    psib50 = object$parameters$psib50,
    sigma = object$parameters$sigma,
    R2 = object$parameters$R2)
}

#' @export
plot.bradford <- function(x, ...) # Plots Bradford's model
{
  colnumber <- length(unique(x$data$treatment))
  colramp <- colorRampPalette(c("red", "orange", "yellow",
                                "green", "blue", "violet"))
  x$data$color <- colramp(colnumber)[as.numeric(cut(x$data$treatment, breaks = colnumber))]

  plot(x$data$psibg, x$data$probit, col = x$data$color, pch = 16,
       xlab = expression(paste(psi[b], " (g)")), ylab = "Probit germination")
  abline(lm(x$data$probit ~ x$data$psibg))
  legend("topleft", title = expression(paste(psi, " (MPa)")),
         legend = levels(as.factor(round(x$data$treatment, 1))), pch = 16,
         col = colramp(colnumber))
}

# bradford internal functions

bradtheta <- function(probit, x, invt) # Calculates Bradford's Theta
{
  finite <- is.finite(probit) & is.finite(x) & is.finite(invt)
  probit <- probit[finite]
  x <- x[finite]
  invt <- invt[finite]

  Sxy <- cov(probit, x)
  Sxz <- cov(probit, invt)
  Sxx <- var(probit)
  Syz <- cov(x, invt)
  Syy <- var(x)
  Szz <- var(invt)

  a <- -2 * (Sxz ^ 2) * Syz + 2 * Sxy * Sxz * Szz
  b <- 2 * (Sxz ^ 2) * Syy + 4 * Sxy * Syz * Sxz - 2 * (Sxy ^ 2) *
    Szz - 4 * Syz * Sxy * Sxz
  c <- 2 * Syz * Sxy ^ 2 - 2 * Sxy * Sxz * Syy

  max(c(( -b + sqrt(b ^ 2 - 4 * a * c))/(2 * a),
        (-b -sqrt(b ^ 2 - 4 * a * c))/(2 * a)))
}
